From 80cb6d37008c0919f6414c190d7c672873747556 Mon Sep 17 00:00:00 2001
From: P Praneesh <quic_ppranees@quicinc.com>
Date: Thu, 17 Nov 2022 10:54:08 +0530
Subject: [PATCH] ath12k: WAR for memory leak caused by thermal register

wiphy_free deallocates rdev via put_device which internally uses
kref_count of kobj. This kref_count gets incremented during device_add and
decremented during device_del. During rmmod, wiphy_free expectes this
kref_count to be zero for deallocating rdev.

Ath12k driver uses devm_hwmon_device_register_with_groups for thermal
register by passing hw->wiphy->dev which is the child of pci->dev. So the
kref_count incremented for the hw->wiphy->dev is not getting decremented
properly during device_del which causes stale value present in the
kref_count. Because of the positive value in kref_count, put_device failed
to call kfree of rdev which causes memory leaks.

Add a workaround to handle this memory leak, by passing pci's dev pointer
instead of wiphy->dev which got created during ieee80211_alloc_hw_nm.

Need to check devm_hwmon_device_register_with_groups internals to fix it
properly.

Signed-off-by: P Praneesh <quic_ppranees@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/thermal.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/drivers/net/wireless/ath/ath12k/thermal.c b/drivers/net/wireless/ath/ath12k/thermal.c
index e29a488..b1d7575 100644
--- a/drivers/net/wireless/ath/ath12k/thermal.c
+++ b/drivers/net/wireless/ath/ath12k/thermal.c
@@ -199,7 +199,7 @@ int ath12k_thermal_register(struct ath12k_base *ab)
 		if (!IS_REACHABLE(CONFIG_HWMON))
 			return 0;
 
-		hwmon_dev = devm_hwmon_device_register_with_groups(&hw->wiphy->dev,
+		hwmon_dev = devm_hwmon_device_register_with_groups(ab->dev,
 								   "ath12k_hwmon", ar,
 								   ath12k_hwmon_groups);
 		if (IS_ERR(hwmon_dev)) {
-- 
2.17.1

